File: clients/ruby/debugging/camera.rb
require './model/vec2'

module Debugging

# Camera state
class Camera
    # Center
    attr_accessor :center
    # Rotation
    attr_accessor :rotation
    # Attack angle
    attr_accessor :attack
    # Vertical field of view
    attr_accessor :fov

    def initialize(center, rotation, attack, fov)
        @center = center
        @rotation = rotation
        @attack = attack
        @fov = fov
    end

    # Read Camera from input stream
    def self.read_from(stream)
        center = Model::Vec2.read_from(stream)
        rotation = stream.read_double()
        attack = stream.read_double()
        fov = stream.read_double()
        Camera.new(center, rotation, attack, fov)
    end

    # Write Camera to output stream
    def write_to(stream)
        @center.write_to(stream)
        stream.write_double(@rotation)
        stream.write_double(@attack)
        stream.write_double(@fov)
    end

    def to_s
        string_result = "Camera { "
        string_result += "center: "
        string_result += @center.to_s
        string_result += ", "
        string_result += "rotation: "
        string_result += @rotation.to_s
        string_result += ", "
        string_result += "attack: "
        string_result += @attack.to_s
        string_result += ", "
        string_result += "fov: "
        string_result += @fov.to_s
        string_result += " }"
        string_result
    end

    def to_str
        to_s
    end
end

end

File: clients/go/pkg/debugging/camera.go
package debugging

import (
	"fmt"
	"io"

	"aicup22/pkg/model"
	"aicup22/pkg/stream"
)

// Camera state
type Camera struct {
	// Center
	Center model.Vec2
	// Rotation
	Rotation float64
	// Attack angle
	Attack float64
	// Vertical field of view
	Fov float64
}

var (
	flow *stream.Stream
)

func NewCamera(center model.Vec2, rotation, attack, fov float64) Camera {
	flow = stream.Flow()
	return Camera{
		Center:   center,
		Rotation: rotation,
		Attack:   attack,
		Fov:      fov,
	}
}

// Read Camera from reader
func ReadCamera(reader io.Reader) Camera {
	center := model.ReadVec2(reader)
	rotation := flow.ReadFloat64()
	attack := flow.ReadFloat64()
	fov := flow.ReadFloat64()
	return Camera{
		Center:   center,
		Rotation: rotation,
		Attack:   attack,
		Fov:      fov,
	}
}

// Write Camera to writer
func (camera Camera) Write(writer io.Writer) {
	center := camera.Center
	center.Write(writer)
	rotation := camera.Rotation
	stream.WriteFloat64(writer, rotation)
	attack := camera.Attack
	stream.WriteFloat64(writer, attack)
	fov := camera.Fov
	stream.WriteFloat64(writer, fov)
}

// Get string representation of Camera
func (camera Camera) String() string {
	stringResult := "{ "
	stringResult += "Center: "
	center := camera.Center
	stringResult += center.String()
	stringResult += ", "
	stringResult += "Rotation: "
	rotation := camera.Rotation
	stringResult += fmt.Sprint(rotation)
	stringResult += ", "
	stringResult += "Attack: "
	attack := camera.Attack
	stringResult += fmt.Sprint(attack)
	stringResult += ", "
	stringResult += "Fov: "
	fov := camera.Fov
	stringResult += fmt.Sprint(fov)
	stringResult += " }"
	return stringResult
}
